plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeIFENotCanonical.sol

```diff
@@ -0,0 +1,125 @@
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+import "../PaymentExitDataModel.sol";
+import "../routers/PaymentInFlightExitRouterArgs.sol";
+import "../spendingConditions/IPaymentSpendingCondition.sol";
+import "../spendingConditions/PaymentSpendingConditionRegistry.sol";
+import "../../utils/ExitId.sol";
+import "../../../utils/UtxoPosLib.sol";
+import "../../../utils/Merkle.sol";
+import "../../../framework/PlasmaFramework.sol";
+
+library PaymentChallengeIFENotCanonical {
+    using UtxoPosLib for UtxoPosLib.UtxoPos;
+
+    struct Controller {
+        PlasmaFramework framework;
+        PaymentSpendingConditionRegistry spendingConditionRegistry;
+    }
+
+    event InFlightExitChallenged(
+        address indexed challenger,
+        bytes32 txHash,
+        uint256 challengeTxPosition
+    );
+
+    function run(
+        Controller memory self,
+        PaymentExitDataModel.InFlightExitMap storage inFlightExitMap,
+        PaymentInFlightExitRouterArgs.ChallengeCanonicityArgs memory args
+    )
+        public
+    {
+        // Check if there is an active in-flight exit from this transaction?
+        uint192 exitId = ExitId.getInFlightExitId(args.inFlightTx);
+        PaymentExitDataModel.InFlightExit storage ife = inFlightExitMap.exits[exitId];
+        require(ife.exitStartTimestamp != 0, "In-fligh exit doesn't exists");
+
+        // Check that the exit is active and in period 1
+        verifyFirstPhaseNotOver(ife, self.framework.minExitPeriod());
+
+        // Check that two transactions are not the same
+        require(
+            keccak256(args.inFlightTx) != keccak256(args.competingTx),
+            "The competitor transaction is the same as transaction in-flight"
+        );
+
+        PaymentTransactionModel.Transaction memory inFlightTx = PaymentTransactionModel.decode(args.inFlightTx);
+
+        // Check that shared input owner signes competing transaction
+        IPaymentSpendingCondition condition = self.spendingConditionRegistry.spendingConditions(
+            args.competingTxInputOutputType, inFlightTx.txType
+        );
+        require(address(condition) != address(0), "Spending condition contract not found");
+
+        bool isSpentByInFlightTx = condition.verify(
+            ife.inputs[args.inFlightTxInputIndex].outputGuard,
+            uint256(0), // should not be used
+            inFlightTx.inputs[args.inFlightTxInputIndex],
+            args.competingTx,
+            args.competingTxInputIndex,
+            args.competingTxWitness
+        );
+        require(isSpentByInFlightTx, "Competing input spending condition is not met");
+
+        // Determine the position of the competing transaction
+        uint256 competitorPosition = ~uint256(0);
```

Comment:
I'd like to add another comment about, the trick used here might not be really intuitive at first glance.
```
// set to infinite large for in-flight competing transaction
```